### PR TITLE
docs: relax required approvals recommendation

### PR DESCRIPTION
**Pull request by tancnle**
# Description

This commit adds some caveats around the required approvals recommendation. We also add some information on how to configure the approvals settings in GitHub repos.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (update or new)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code

### CODE REVIEW

File: docs/processes/quality-assurance/git-contribution-guide.md
Comment (ginge270) [inline docs/processes/quality-assurance/git-contribution-guide.md]: I think rather than saying please try to have another person look at the changes, it should say both approvals should be people other than the author. Also clarifying who the second approval should be?

Comment (ginge270) [reply docs/processes/quality-assurance/git-contribution-guide.md]: eg. Area Lead

Comment (tancnle) [reply docs/processes/quality-assurance/git-contribution-guide.md]: Thanks, @ginge270 

> I think rather than saying please try to have another person look at the changes, it should say both approvals should be people other than the author.

My intention is not to enforce a minimum of 2 approvals but to let teams decide to balance velocity and quality. For example, a small team might just need one or the delivery lead might not involve with much of the technical details. Furthermore, the system will not allow authors to approve themselves. Another intricacy is work done by a pair. Should the team allow it to expedite the approval process? I lean towards no but teams might have different opinions.

> Area Lead

This is an interesting one. I think approval should stay with the knowledge authority, not hierarchical authority. The process is to ensure we collect relevant insights and innovate, but less about asking for permission to execute. For example, some might not know about C++ or build systems so PRs around this area might not be suited to direct to them. It is good for upskilling as FYI but not from approval standpoint.

Comment (ginge270) [reply docs/processes/quality-assurance/git-contribution-guide.md]: I think this should be discussed further as my understanding from our QA meetings is not aligned with this.

An area lead or product lead may not have the knowledge authority, but QA is not just about the coding itself. I'm not making a comment about hierarchy - but right now we are the ones most familiar with this vision and part of teaching this will be through feedback. Having someone to look at the PR to ensure that it meets the standards around readability, understandability, reusability - not just functionality and these are all transferable skills from other languages. Has the person who has submitted the PR proven that this code does what it says it will? Is someone inexperienced with the language able to get a sense of what it is doing based on the variable and function names? Does it make sense? A peer or member of their team will be looking at it from a similar lens with the same context. 
 
 "Please try to have at least another person to look at the changes beside the author" implies that the system will allow a merge without approval or the author can approve themself - this is why I made the comment about being more explicit.

Comment (tancnle) [reply docs/processes/quality-assurance/git-contribution-guide.md]: Thank you for elaborating on your thoughts ❤️ I think we are aligned in principle. 

> knowledge authority, but QA is not just about the coding itself.

Yes, and knowledge authority is not about coding only.

> we are the ones most familiar with this vision and part of teaching this will be through feedback.

Yes totally, and I agree this is driven by knowledge, not authority.

> Having someone to look at the PR to ensure that it meets the standards around readability, understandability, reusability - not just functionality and these are all transferable skills from other languages...Does it make sense? A peer or member of their team will be looking at it from a similar lens with the same context.

This is a great point and I think it captures the intention of the approval process really well. We can include that in the doc. To be clear, I am fully supportive to expose the PR to more parties to invite diverse opinions, but not forcing approvals if it does not make sense. That needs to balance with velocity, everyone is working async and if we have a lot of required approvals, that would slow down the process and potentially affect people who propose changes. 

> "Please try to have at least another person to look at the changes beside the author" 

Sorry if it sounds confusing. I can try to reword it.

Comment (ghost) [reply docs/processes/quality-assurance/git-contribution-guide.md]: I just thought I would add my untechnical opinion.
Getting two approvals is difficult, for this handbook we have had to contact each other numerous times just to have one approved review.

I understand the importance but with small teams this might be more difficult than we assume. Having code sit waiting for someone to approve it will drastically increase stress levels, create merging issues, and potentially reduce the actual care around quality - e.g. people approving anything.

Or are we assuming that by publishing the PR this is counted as 1 review?

Comment (kats8) [reply docs/processes/quality-assurance/git-contribution-guide.md]: I have a suspicion practicality might force us to have a balance. Although... docs might be the kind of thing that you'd expect to end up with more PRs and less approvals needed, so maybe not representative. I think my tendency is to overload my coding commits/PRs so my gut feel on that might not be great either.

In my original vision (when we had different branch structures and PRs-getting-to-Doubtfire methods proposed) I was thinking the PRs would be much bigger chunks that "higher level" leads might be approving (with a much higher level, superficial kind of assessment than, say, @Tan usually does).

For instance, in my mind, this was where the "folder/branch for each project" structure came in - eg, a coder might have another reviewer approve their sub-branch to project branch merge, but then PRs to periodically merge changes to main would have a higher-level approver (so you'd need an area lead to look it over before it got into Thoth-Tech main, but there might have been many smaller steps that were peer-reviewed to get to the project branch before it was merged).

I now understand there's reasons for smaller PRs being approved & merged (including increasing potential for merge issues) but having removed the interim project branches from the flow, it does mean rather than fewer (more manageable) "check-point" PRs to assess (with larger, less manageable(?) code chunks), we may run into a pure PR quantity issue, as @talzid mentions.

We could have a "training period" where we force the extra approval to try to force quality standards 
until confidence is built with teams that they've got the hang of it and/or delegate this further down to delivery leads as they show they're confident in QA control.

- problem I see there is that this is the training period for everything and I can imagine this confidence might take a bit of the remaining trimester.
- of course, the other side of that coin is that if there's reason for that confidence to be shaky now, the chance of QA being followed without some approval gate-keeping may not be so good?

If, hypothetically, we think it's coming down to (A) Getting appreciable coding done (with 1 peer review only) or (B). Slowing progress to a trickle (but having more solid QA) could there be an option (C) whereby we let them go with 1 approval but do auditing instead and feed that back as issues? (I don't know much about issues - can they be raised post-merged-PR looking at prior changes?)

Comment (tancnle) [reply docs/processes/quality-assurance/git-contribution-guide.md]: @talzid Thank you for sharing your thoughts 🙏🏼 

> Getting two approvals is difficult

Yep. I share your concern. It's hard for small teams and/or depending on what stage our software product is. I would like to encourage teams to look at the suggestion and within their own capacity, reason about the appropriate approval process for them. 

Another thing is trying to balance individual workload with the expectation around the review cycle, (e.g. PR should receive a response in 24 hours). We can provide some more guidance on this one in follow-up PR.

> Or are we assuming that by publishing the PR this is counted as 1 review?

I don't think publishing the PR is counted as 1 review. A PR can have many reviewers but only receive approval from some. Or might be I misunderstood your question completely 🤔

File: docs/processes/quality-assurance/git-contribution-guide.md
Comment (ghost) [inline docs/processes/quality-assurance/git-contribution-guide.md:180]: I do like how this refers to a domain expert - I think this provides the clarity we are after surrounding reviewers. This also addresses team size issues.

I think once the people start getting engaged with PRs and GitHub as a whole we will see a happy medium with quality and velocity. I also think this will tip toward velocity closer to the deadline.
As long as we raise awareness about our expectations this should improve throughout the companies life span. 
- highlighting good examples
- showing areas of improvement
- ect.

We want to empower our employees, thus we should also trust they will educate themselves accordingly

File: docs/processes/quality-assurance/git-contribution-guide.md
Comment (tancnle) [reply docs/processes/quality-assurance/git-contribution-guide.md]: Thank you for your input @kats8 👍🏼 

>  I have suspicion practicality might force us to have a balance

Yes, it is. I would like to massage the message to give people that flexibility.

> We could have a "training period" where we force the extra approval to try to force quality standards

That's a great idea and in fact, most open-source teams work that way. Practically, I am not sure how we can do that given the limited time of the course and concentration of knowledge across products and technologies.

> (A) Getting appreciable coding done (with 1 peer review only) 
> (B) Slowing progress to a trickle (but having more solid QA)
> (C) whereby we let them go with 1 approval but do auditing instead and feed that back as issues

I think teams should go with 2 as recommended. Observe the quality metrics (bugs, incidents) and readjust. C is subjected to availability (when you said we, I assumed it is the Area Leads). The team can even do that themselves, for example, run a bug bash day at the end of the iteration.

> I don't know much about issues - can they be raised post-merged-PR looking at prior changes?

Yep, we can raise bug issues on the Trello board or PRs (if straight-forward and urgent) to address problems introduced by previous changes.

File: docs/processes/quality-assurance/git-contribution-guide.md
Comment (tancnle) [reply docs/processes/quality-assurance/git-contribution-guide.md:180]: Thank you for the feedback @talzid. 

> I do like how this refers to a domain expert 

I have some ideas to bring the skill matrix -> domain expert -> review automation into action :)

> We want to empower our employees, thus we should also trust they will educate themselves accordingly

Well put 👏🏼

File: docs/processes/quality-assurance/git-contribution-guide.md
Comment (ginge270) [reply docs/processes/quality-assurance/git-contribution-guide.md]: I think teams should go with 2 as recommended. Observe the quality metrics (bugs, incidents) and readjust. C is subjected to availability (when you said we, I assumed it is the Area Leads). The team can even do that themselves, for example, run a bug bash day at the end of the iteration.

I agree with this outcome!

Comment (ginge270) [reply docs/processes/quality-assurance/git-contribution-guide.md]: Also, on the point that Katrina brought up - I think it should be clear in the guidelines that PRs should not be big pieces of work. They should be smaller chunks of functional work. As well as commit often

Comment (tancnle) [reply docs/processes/quality-assurance/git-contribution-guide.md]: > Also, on the point that Katrina brought up - I think it should be clear in the guidelines that PRs should not be big pieces of work. They should be smaller chunks of functional work. As well as commit often

In the spirit of iteration and small PR, I will create a follow-up PR to include this as a sub-section. Potentially include the content from the Git/GitHub training [slide](https://deakin365.sharepoint.com/:p:/r/sites/ThothTech2/_layouts/15/Doc.aspx?sourcedoc=%7BDE2764FD-E00C-42C5-BAA1-A2D7DDA4F444%7D&file=Git%20and%20GitHub%20Training.pptx&action=edit&mobileredirect=true). What do you think?

Comment (ghost) [reply docs/processes/quality-assurance/git-contribution-guide.md]: If we are wanting to create some more work... we could turn the Github training session into .md and put it in training and link to that.

Comment (tancnle) [reply docs/processes/quality-assurance/git-contribution-guide.md]: Yep. It's on my list :)

**Review by ginge270 — APPROVED** [review]

**Pull request merged**
